[Dynamo] Fix a corner case of reinplace_inplaceable_ops pass for triton kernels

sijiac opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #117612

Summary:
We saw the following failure when compiling custom triton kernels:
```
RuntimeError: Argument 'getitem_22' of Node 'triton_kernel_wrapper_functional_proxy_3' was used before it has been defined! Please check that Nodes in the graph are topologically ordered
```
The root-cause is when doing the replacement, the replacement is replaced by another replacement. The fix will keep finding the replacement until it is not replaced

Test Plan:

Added a test case

Reviewers:

Subscribers:

Tasks:

Tags:

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

sijiac committed:
[Dynamo] Fix a corner case of reinplace_inplaceable_ops pass for triton kernels

Summary:
We saw the following failure when compiling custom triton kernels:
```
RuntimeError: Argument 'getitem_22' of Node 'triton_kernel_wrapper_functional_proxy_3' was used before it has been defined! Please check that Nodes in the graph are topologically ordered
```
The root-cause is when doing the replacement, the replacement is replaced by another replacement. The fix will keep finding the replacement until it is not replaced

Test Plan:

Added a test case

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

aakhundov commented on torch/_inductor/fx_passes/post_grad.py:
```diff
@@ -795,8 +795,16 @@ def can_inplace(node, mutated_arg):
-    for node, replacement in replace_list:
-        node.replace_all_uses_with(replacement)
+
+    def get_replacement(node):
+        ori_node = node
+        while node in replace_dict:
+            node = replace_dict[node]
+        replace_dict[ori_node] = node
+        return node
```
Nit: as this is done only once for every distinct key node in the `replace_dict`, I'm wondering if we should just inline this in the for-loop below. Then also could start from the `replace_dict[node]` to save one iteration and skip the assignment as the same `replace_dict[node]` won't be queried again.

aakhundov commented on torch/_inductor/fx_passes/post_grad.py:
```diff
@@ -795,8 +795,16 @@ def can_inplace(node, mutated_arg):
+
+    def get_replacement(node):
+        ori_node = node
+        while node in replace_dict:
+            node = replace_dict[node]
```
@oulgen To make sure: we can't have cycles in the `replace_dict`, right? I assume it is a DAG.

oulgen replied:
nope, it would require fx graph to have a cycle which is not possible

aakhundov reviewed (approved):
Thanks for the fix!

oulgen commented on torch/_inductor/fx_passes/post_grad.py:
```diff
@@ -795,8 +795,16 @@ def can_inplace(node, mutated_arg):
-    for node, replacement in replace_list:
-        node.replace_all_uses_with(replacement)
+
+    def get_replacement(node):
```
this is called from one place, maybe just inline it, doesn't need to be a helper function

sijiac committed:
Update on "[Dynamo] Fix a corner case of reinplace_inplaceable_ops pass for triton kernels"


Summary:
We saw the following failure when compiling custom triton kernels:
```
RuntimeError: Argument 'getitem_22' of Node 'triton_kernel_wrapper_functional_proxy_3' was used before it has been defined! Please check that Nodes in the graph are topologically ordered
```
The root-cause is when doing the replacement, the replacement is replaced by another replacement. The fix will keep finding the replacement until it is not replaced

Test Plan:

Added a test case

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

oulgen commented on torch/_inductor/fx_passes/post_grad.py:
```diff
@@ -795,7 +795,12 @@ def can_inplace(node, mutated_arg):
+    for node, replacement in replace_dict.items():
+        while replacement in replace_dict:
+            replacement = replace_dict[replacement]
+        replace_dict[node] = replacement
```
It is okay to modify the dict on existing keys while iterating

sijiac commented:
@pytorchbot merge
